rfc/0075-faster-failover-and-configuration-push.md

```diff
@@ -0,0 +1,315 @@
+    Extra length (4)    : 0x10 (two int64_t fields in extras)
+    Data type    (5)    : 0x00 (RAW)
+    Vbucket      (6,7)  : 0x0000
+    Total body   (8-11) : 0x00000010 (16 bytes)
+    Opaque       (12-15): 0x00000000
+    CAS          (16-23): 0x0000000000000000
+    Epoch        (24-31): 0x0000000000000042 (66 in base-10)
+    Revision     (32-39): 0x0102030405060708 (72623859790382856 in base-10)
+
+So note that magic is `ServerRequest` (`0x82`), that is enabled by `Duplex` (`0x0c`) HELLO flag. Also note that just
+like in regular cluster configuration notification, epoch and revision fields are sent as extras.
+
+Note that the magic value for this notification is `ServerRequest` (`0x82`), which is enabled by the `Duplex` (`0x0c`)
+HELLO flag. Additionally, similar to the regular cluster configuration notification, the epoch and revision fields are
+sent as extras.
+
+Once the brief cluster configuration notification is received, it is up to the SDK to decide whether to send a
+`GetClusterConfig` (`0xb5`) request to retrieve the actual configuration body.
+
+In essence, the `ClustermapChangeNotificationBrief` feature only saves network traffic. If
+`DedupeNotMyVbucketClustermap` is not enabled, the number of notifications will be the same as before. However, this
+feature can still be used as a building block to implement a debouncing mechanism. When properly configured, it can help
+reduce the number of requests. Further details on this topic will be covered in the "Library Changes" section.
+
+## Library Changes
+
+### Configuration Push
+
+The previously mentioned `ClustermapChangeNotificationBrief` feature enables the SDK to subscribe all connections for
+configuration updates. These notifications are lightweight and can be deduplicated by the server when the
+`DedupeNotMyVbucketClustermap` option is negotiated.
+
+#### Mixed Clusters
+
+In clusters where there is a mix of nodes with older server versions, meaning that some nodes do not acknowledge
+`ClustermapChangeNotificationBrief`, the respective connection should notify the configuration monitor about its lack of
+support for configuration pushes from the server. As a result, the monitor should utilize the old polling mechanism for
+this particular node instead.
+
+### Enhancements in Handling the `NotMyVbucket` Status
+
+Combination of `DedupeNotMyVbucketClustermap` and `ClustermapChangeNotificationBrief` allows to save traffic by not
+sending configuration, if SDK already seen the same revision, and also sends only pair of `Epoch`/`Revision`. So it is
+up to SDK to initiate configuration update once the non-empty payload returned along with `NotMyVbucket` status code.
+
+Several modifications are required in the SDK:
+1. The retry orchestrator should be able to retry an operation based on configuration updates rather than the timer signal.
+2. The configuration monitor should have the ability to throttle configuration requests due to the following reasons:
```

Comment:
What is meant by throttle configuration requests here? Only allow a single configuration request to be in flight at once? If that per node or across the cluster?

Reply:
yes, one get_config request in-flight, but also it might delay sending the request by adding some extra time to wait before sending request, so that it will absorb more config update requests in big clusters.

Reply:
I checked that `NotMyVbucket` actually renders full configuration (if it hasn't been seen by SDK previously), so this paragraph does not apply anymore. The section text is updated.

https://github.com/couchbase/kv_engine/blob/0a84f5b4f2377c9c6553e83091d0ab1b527cfd28/daemon/cookie.cc#L267-L300